### ISSUE TITLE
Use Python's logging ability and remove all print statement 

### ISSUE THREAD

**Issue by eprochasson** [sassoftware/saspy#399]
**Is your feature request related to a problem? Please describe.**

When using the saspy module, the output of my script is corrupted by saspy print statements peppered all over the code. A well behaved Python module should not directly write to STDOUT through `print`.

**Describe the solution you'd like**
Replace all print statement with calls to the logging module. Each file that want to emit some kind of message should

- import the logging module `import logging`
- instantiate a default logging session, e.g.: `logger = logging.getLogger(__name__)`. This will automatically make the name of the current file available to the logging module. When calling it, the users will be able to see where the message was emitted from (e.g.: `saspy.sasioiom`), and also will give the user the ability to silence the logging.
- replace the print statement with proper logging calls. For example, in `sasbase.py`, around L570:

```python
        if self.sascfg.mode != 'HTTP':
           try:
              self.pyenc = sas_encoding_mapping[self.sascei]
           except KeyError:
              logger.fatal("Invalid response from SAS on inital submission. printing the SASLOG as diagnostic")
              logger.fatal(self._io._log)
              raise
```

The logging module makes it easy to indicate what's debug / info / warning / critical information (which is a lot more flexible than the simple "verbose" mode -- user can simply chose to show messages down to a certain level).

**Describe alternatives you've considered**

I'm happy to propose a PR regarding this issue, as it's fairly trivial, but would like to see if it's something you'd be interested in.

**Comment by tomweber-sas** [comment]
@eprochasson, yes, I agree that this is a much better way than using print statements. I have to admit the print statements have been in here since day 1 when I was teaching myself Python to write this. I agree it's time to clean that up and use the better method, via the logging facility. If you'd like to do a PR for this, that would be great. If not, I'm open to working through it myself, as I think this would be a useful enhancement. Thanks for proposing this, I think it will be a good thing to implement!
Tom

**Comment by tomweber-sas** [comment]
So, I've been exploring this (I haven't used logging before), and this is pretty cool. Very straight forward and I should have done this a long time ago :)
So, I do have an implementation (partial, just a couple parts of the whole system using it so far), which is doing what I expect from this. Though, I'm not sure it's the same as what your idea of it would be. I have some constraints, like backward compatibility, that I have to worry about. So changing what all of the messages are, as opposed to being able to control which are displayed, is something that might not jive completely. But, let's see.

What I've done just today is to create a logger for saspy. With this, I'm defaulting the level to INFO. Most all (I only made 1 print debug) existing messages will be displayed with this, same as they are displayed currently (backward compatible, and not breaking anyone's code). But, 'you' (the user) can simply configure the logger at will to have it do what they want; change the logging level, route it to a file, whatever you feel like. One thing this doesn't have is a logger per .py file in the whole saspy module. I don't think most users of this have any idea what files are what or care which file a message came from; it's just saspy and the method they ran that produces the message. We can talk about that though.

So, here's what I've coded up so far, as my first attempt to use this. After importing saspy, saspy.logger is accessible to be adjusted. Level is INFO, but you can just change that as you like. Propagation of turned off, as that changes the messages from as they are now to as they are, plus another line with the LEVEL:saspy:msg printed out, redundantly (but you can change it to True if you like):
```
>>> sas.sil(3,3)
Must suply 3 of the 4 variables. Only 2 variables provided.
ERROR:saspy:Must suply 3 of the 4 variables. Only 2 variables provided.
>>>

instead of
>>> sas.sil(3,3)
Must suply 3 of the 4 variables. Only 2 variables provided.
>>>
```
Here's a little example of changing the logging levels, and I'm sure you can swap out the handlers to route to a file instead of the console, so that should give you plenty of control to have it the way you want, without changing anything by default (breaking how it already works).
```
tom64-5> python3.5
Python 3.5.6 (default, Nov 16 2018, 15:50:39)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-23)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import saspy
>>> sas = saspy.SASsession()
Using SAS Config named: iomj
SAS Connection established. Subprocess id is 9769

>>> sd = sas.sasdata('bob')
Table WORK.bob does not exist. This SASdata object will not be useful until the data set is created.
>>> sas.sil(34,34)
Must suply 3 of the 4 variables. Only 2 variables provided.
>>> saspy.logger.setLevel(40)
>>> sas = saspy.SASsession()
>>> sd = sas.sasdata('bob')
>>> sd = sas.sasdata('bob')
>>> sas.sil(34,34)
Must suply 3 of the 4 variables. Only 2 variables provided.
>>> saspy.logger.setLevel(50)
>>> sas.sil(34,34)
>>> sas.endsas()
>>>
>>>
tom64-5>
```
I've pushed what I have so far to a new branch called ... wait for it ... 'logging' :)  I'll continue to go though more of the code and change it over to this and push there. It's pretty pervasive, but easy to change, so I'll push as I get each piece done. Give it a try and see what you think. 

Thanks!
Tom

**Comment by tomweber-sas** [comment]
I have my 3 sasio* access methods, along with sasbase and sasdata converted. So that's a lot to be able to play with and see how it behaves. I did have to change the default output destination from stderr to stdout (which is where the prints went so still backward compatible), because Jupyter makes everything written to stderr be RED highlighted, so that's not how it ever looked before and too different, especially for just INFO messages. Now what's there is looking the same as before, but you can (re)configure it to be as you like for your program and use of saspy.

Again, give it a go and see what you think.

Thanks!
Tom

**Comment by tomweber-sas** [comment]
FYI, I finished the changes in the rest of the modules files. So everything is pushed into this branch.
Tom

**Comment by tomweber-sas** [comment]
Here's a pic (hope it resolves) showing some of the manipulation you can do (any really) with the saspy logger, including showing jupyter using RED for info msgs when routed to stderr as opposed of my default of stdout (see the 'SAS Connection established. Subprocess id is nnnnnn' msgs) . But also routing to a file, and you can reconfigure the logger at will.

![image](https://user-images.githubusercontent.com/17710182/130122426-659c27f9-ddd7-4bc6-b212-1b0ff06820f4.png)

**Comment by tomweber-sas** [comment]
Had a chance to look at any of this? If it's good to you, I can merge it in and it'll then be in the next release.

Thanks,
Tom

**Comment by eprochasson** [comment]
Hi,

first, apologies are due for the late reply. FYI, I use SAS for work, work which does not allow me access to Github, I can only look into this outside of work.

The bulk of the work looks correct to me: the most important (if not all) print statement have been effectively replaced.

However, in the `__init__.py` file, you should not specify anything beyond the logger name. A module should not force logging to STDOUT or at a specific level. It's the responsibility of the user of the module to decide that.

By default, Python will show anything above and including "WARNING" level to STDOUT. Meaning this will still work as in your example, in Jupyter notebook, if you don't specify anything, while letting the user decide otherwise.

Many thanks! And thanks for all your work on this module, it does save my life at work.

**Comment by tomweber-sas** [comment]
Hey @eprochasson, yeah, no problem, I know lots of places have gihub restricted from work computers, so I understand.

As for tweaking the defaults for the SASPy implementation, as opposed to using Python's defaults, that's a matter of backward compatibility. Had I done it this way 6 years ago, using logging instead of print(), then I would have just let the defaults be whatever they happened to be. But, since I'm rearchitecting this now, having it be backwards compatible is more important to me. Any new things I create, I'll use logging and the defaults, so I won't have this issue.

Like you say, anyone can change the behavior if they like, and that's still true. In fact, when I push this out, I'll be documenting it, and will show examples of how to modify it to your liking (including the few lines to make if be python default), so anyone who actually wants it that way can simple add that, or whatever configurations they want, to their config file and it will just be that way for them all the time, by default.

But, I can't break existing code, even if that just means every message all of a sudden shows up with red reverse video highlighting in their notebooks. Because, that's not how it works today and is too much of a significant change in every notebook, and that's not a backward compatible change. Not to mention that most of the notes will go away since the default it WARNING, not NOTE. So, backward compatibility trumps defaults when you can simply set it to be any way you like (which you can do).

I believe all you need in your config to make it always be the python default is this; is that true?
```
import logging
saspy.logger.handlers  = []
saspy.logger.level     = (logging.WARNING)
saspy.logger.propagate = True
```
Can you verify that setting that in your config file makes it be the python defaults you'd like? I think so, but you can confirm.

Thanks!
Tom

**Comment by tomweber-sas** [comment]
Hey, this is now in the current production version, v3.7.5, out on pypi and in the repo. Will build on Conda in a while.
I'm going to go ahead and close this, but if there's anything else, feel free to reopen or open a new one. Thanks again for requesting this, I do like it way better than the print() statements I had scattered all over the place!

Thanks,
Tom
